### ISSUE TITLE
Is there a good reason why T_mult should be an integer

### ISSUE THREAD

**Issue by JulienMaille** [katsura-jp/pytorch-cosine-annealing-with-warmup#4]
Thanks for your work!

**Comment by JulienMaille** [comment]
Here is my own implementation if someone needs it

```python
from torch.optim.lr_scheduler import CosineAnnealingWarmRestarts

class CosineAnnealingWarmUpRestarts(CosineAnnealingWarmRestarts):
    def __init__(self, optimizer, T_0, T_mult=1, eta_min=0, last_epoch=-1, T_up=0, gamma=1.):
        self.gamma = gamma
        self.weight = 1.
        self.T_up = self.T_up_i = T_up
        super().__init__(optimizer, T_0, T_mult=1, eta_min=eta_min, last_epoch=last_epoch)
        self.T_mult = T_mult

    def get_lr(self):
        if self.T_up_i < self.T_up:
          return [(base_lr * self.weight)*self.T_up_i / self.T_up for base_lr in super().get_lr()]
        else:
          return [base_lr * self.weight for base_lr in super().get_lr()]

    def step(self, epoch=None):
        if hasattr(self, 'T_cur') and self.T_cur+1 >= self.T_i:
            self.T_up_i = 0
            self.weight *= self.gamma

        if 0 <= self.T_up_i < self.T_up:
          self.T_up_i += 1

        super().step(epoch=epoch)
```
`scheduler = CosineAnnealingWarmUpRestarts(optimizer, T_0=100, T_mult=1.5, T_up=10, gamma=0.5)`
![image](https://user-images.githubusercontent.com/182520/101982247-c987f180-3c72-11eb-97b1-318c059ea209.png)

**Comment by katsura-jp** [comment]
@JulienMaille 
Thanks for your comment!
I plan to update this implementation in the near future, so I'm going to use your implementation as an example.

thank you.
